_posts/2016-03-16-code-smells-part-I.md

```diff
@@ -0,0 +1,258 @@
+        href: /assets/img/custom/blog/code-smells.jpg
+tags:
+- bloaters
+- code smells
+- long classes/methods
+---
+Last weekend I was at the [SoCraTes Canaries](https://twitter.com/hashtag/socracan16) and I gave my first talk ever about code smells.Oh boy! How nervous was I! But now that that has passed I was wondering what I should do with all information I collected. And then I thought, maybe it's a good idea to put it all in a nice blog post.
+
+## So what are code smells?
+
+As [Martin Fowler](http://martinfowler.com/) said in his book ["Refactoring: Improving the Design of Existing Code"](http://martinfowler.com/books/refactoring.html),
+> **A code smell is a surface indication that usually corresponds to a deeper problem in the system.**
+
+I like to think that a  code smell is something that makes your developer instinct cry out to you, and you just know that something is wrong. This doesn’t mean you have to make changes in your code: there are occasions where these code smells are ok, but I think It’s important for us to detect them and know exactly why they are there.
+
+There are 5 categories in code smells:
+
+- Bloaters
+- Object-Orientation Abusers
+- Change Preventers
+- Dispensables
+- Couplers
+
+Today I'm going to talk about Bloaters. I'll leave the other categories for a future post.
+
+## Bloaters
+
+Bloaters can be big methods or classes, primitive obsessions, data clumps, or long parameters list.
+
+#### Long Parameter List/Data Clumps
+
+The Long Parameter List is when you have a method that has more than 3 parameters. Sometimes we see that when we receive an object, and instead of passing it all we pass some of its data. In this case, the best policy is to pass the whole object. Data Clumps are a bit different: they are, in general, primitive values that start do "get together". A good example of this is a startDate and endDate... Maybe it's worth creating a DateRange.
+
+#### Primitive Obsession
+
+This case is when we use primitives instead of small objects for simple tasks. Sometimes the use of primitives is justifiable, but when you start to have behaviour attached to this primitives, then it's time to stop and think that maybe a value type is in order. A simple example is a currency: we tend to put it in a float/double, instead of encapsulating it in a value type.
+
+#### Long Method / Large Class
+This kind of code smell happens when you have a big method. But, when do you know that a method has become too big? Well, I have the rule that with more than 5 lines, you should, at least, look at it again. But, as [Sandro](https://twitter.com/sandromancuso) told me before, the right number of lines are just enough lines so a method only does one thing (and so it conforms to the 1st principle of [SOLID](https://en.wikipedia.org/wiki/SOLID_(object-oriented_design)) the [Single responsibility principle](https://en.wikipedia.org/wiki/Single_responsibility_principle)).
```

Comment:
Suggestion: maybe mention Composed methods, since you already touched this topic in a bit more detail.